### DISCUSSION TITLE
Typescript best practices & HOC

### DISCUSSION THREAD

**Discussion by Ennoriel** [react-hook-form/react-hook-form#7192] [Q&A]
Hi,

This library is a must in react form handling! Thanks for your work!

In my team we are going to use it with Typescript (migrating from redux-form, removing all th redux logic at first).

Our code base is currently made of:
  - input components
  - for each input, a wrapper component that define the redux-form Field
  - a wrapper component that handles label & error display

I would like to keep this structure (I'll come back to shifting point 2 with a HOC at the end).

I am not sure if I need to redefine all the times the generic types or if there is a better solution to achieve what we want.

Here is a MVP code (also available in a working MVP github repo [here](https://github.com/Ennoriel/react-hook-form-ts)):

Basic Text Input
```
const TextInput = (props: React.ComponentPropsWithoutRef<'input'>) => {
    return <input type="text" {...props} />;
};
```

Our magic Field renderer which keeps all label & error the same
```
const FieldRenderer = <TFieldValues extends FieldValues = FieldValues, TName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>>(props: FieldRendererProps<TFieldValues, TName>) => {

    const {
        field,
        fieldState: {error}
    } = useController({
        name: props.name,
        control: props.control,
        rules: props.rules
    });

    return (
        <>
            {props.label}
            {props.renderFunction(field)}
			<pre>{JSON.stringify(error, null, 2)}</pre>
        </>
    );
};
```

The Input wrapped with the fieldRenderer
```
const TextInputWrapper = <TFieldValues extends FieldValues = FieldValues, TName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>>(props: FieldProps<TFieldValues, TName>) => {

    return (
        <FieldRenderer
            {...props}
            renderFunction={TextInput}
        />
    );
};
```

Types
```
export interface ExternalFieldRenderProps {
    label?: string;
}

export interface HookFormProps<TFieldValues extends FieldValues = FieldValues, TName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>> {
    name: TName;
    control: Control<TFieldValues>;
    rules?: RegisterOptions
}

export interface RendererProps<TFieldValues extends FieldValues = FieldValues, TName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>> {
    renderFunction: (props: ControllerRenderProps<TFieldValues, TName> ) => JSX.Element
}

export type FieldRendererProps<TFieldValues extends FieldValues = FieldValues, TName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>> =
    ExternalFieldRenderProps &
    HookFormProps<TFieldValues, TName> &
    RendererProps<TFieldValues, TName>

export type FieldProps<TFieldValues extends FieldValues = FieldValues, TName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>> =
    ExternalFieldRenderProps &
    HookFormProps<TFieldValues, TName>
```

I am not sure if I need to redefine all the times the generic types or if there is a better solution to achieve what we want.

Also, we came up with a HOC but we have type assertion problems (control type is not well asserted) and don't know how to handle the errors here.

```
const FormHoc = <TFieldValues extends FieldValues = FieldValues, TName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>>(Component: React.ComponentType<InputTextProps<TFieldValues, TName>>) =>
    class ShadowClassName extends React.Component<InputTextProps<TFieldValues, TName> & Label> {
        render() {
            const {label, ...props} = this.props;

            return (
                <>
                    {label}
                    <Component {...props} />
                </>
            );
        }
    }
```

Could you help us with that?
Thank you!

**Comment by lundn** [comment]
I hope to evolve my answer during the day, but have you looked into helper types such as `UseControllerProps`?

**Comment by Ennoriel** [comment]
I just looked at it.

I could replace in my type definitions:  HookFormProps by UseControllerProps. They are equivalent. It does not remove the need to write the generic definition for each component. Is there a way to not write it? Maybe by defining the components another way?

```
<TFieldValues extends FieldValues = FieldValues, TName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>>
```

**Reply by lundn** [reply]
I forked your Github repo. Could you have a look and see this gives you any more answers? 
https://codesandbox.io/s/epic-hermann-gdim7?file=/src/App.tsx

**Reply by Ennoriel** [reply]
Thanks for the reply and codesandbox!

Did you change anything else except in Input.tsx?

Locally (either by modifying my project or downloading the code sandbox), I get this ts error that I partially understand in TextInputWrapper, `renderFunction={TextInput}` :

```
Type '(props: ComponentPropsWithoutRef<'input'>) => JSX.Element' is not assignable to type '(props: ControllerRenderProps<T, Path<T>>) => Element'.
  Types of parameters 'props' and 'props' are incompatible.
    Type 'ControllerRenderProps<T, Path<T>>' is not assignable to type 'Pick<DetailedHTMLProps<InputHTMLAttributes<HTMLInputElement>, HTMLInputElement>, "key" | keyof InputHTMLAttributes<...>>'.
      Types of property 'value' are incompatible.
        Type 'UnpackNestedValue<PathValue<T, Path<T>>>' is not assignable to type 'string | number | readonly string[] | undefined'.
The error continues
```

**Reply by Ennoriel** [reply]
Hi @lundn 

Thanks again for your help!

I'd like to update this thread. I haven't found the solution and I am still defining my input components with a props dependening on 2 Generics with 2 default values.

I don't understand why it's working in the codepen editor and not locally. My Ts version is the same.

Have a nice day.

**Reply by Ennoriel** [reply]
Hi @lundn,

I came up with a trick in order to overcome this typescript error:
https://codesandbox.io/s/kind-dhawan-ick4i

```
export function TextInputWrapper<T>(props: FieldProps<T>) {
  return (
    <FieldRenderer
      {...props}
      renderFunction={({ value, ...field }) => (
        <TextInput {...field} value={value as string} />
      )}
    />
  );
}
```

I understand that the library allows all types for value, and not just string in this case. But do you know if there is a simpler solution than to assert the value type?

Thank you for taking your time.
